sample-std/src/lib.rs

```diff
@@ -207,6 +207,102 @@ where
     }
 }
 
+macro_rules! replace_expr {
+    ($_t:tt $sub:expr) => {
+        $sub
+    };
+}
+
+macro_rules! none_pad {
+    (($cur:ident) $($post: ident)*) => {
+        (Some($cur), $(replace_expr!($post  None)),*)
+    };
+    ($($pre:ident)+ ($cur:ident) $($post: ident)*) => {
+        ($(replace_expr!($pre  None)),*, Some($cur), $(replace_expr!($post  None)),*)
+    };
+}
+
+macro_rules! shrink_tuple {
+    ($v:ident () $($sample: ident)*) => {
+        let ($(casey::lower!($sample)),*) = $v.clone();
+        let r = std::iter::empty();
+        shrink_tuple!(r $v () ($($sample)*));
+    };
+    ($r:ident $v:ident () ($cur:ident $($sample: ident)*)) => {
+        let (casey::lower!(ref $cur), $(casey::lower!(ref $sample)),*) = $v;
+        let r = $r.chain(
+            $cur.shrink(casey::lower!($cur).clone()).map(move |$cur| {
+                none_pad!(($cur) $($sample)*)
+            })
+        );
+        shrink_tuple!(r $v ($cur) ($($sample)*));
+    };
+    ($r:ident $v:ident ($($pre:ident)+) ($cur:ident $($sample: ident)*)) => {
+        let ($(casey::lower!(ref $pre)),*, casey::lower!(ref $cur), $(casey::lower!(ref $sample)),*) = $v;
+        let r = $r.chain(
+            $cur.shrink(casey::lower!($cur).clone()).map(move |$cur| {
+                none_pad!($($pre)* ($cur) $($sample)*)
+            })
+        );
+        shrink_tuple!(r $v ($($pre)* $cur) ($($sample)*));
+    };
+    ($r:ident $v:ident ($($pre:ident)*) ()) => {
+        let ($(casey::lower!($pre)),*) = $v;
+        return Box::new($r.map(move |($($pre),*)| {
+            ($($pre.unwrap_or(casey::lower!($pre).clone())),*)
+        }))
+    };
+}
+
+macro_rules! sample_tuple {
+    ($($name: ident),*) => {
+
+impl<$($name),*> Sample for ($($name),*,)
```

Comment:
I'm assuming fmt wants it like that?
Even though I myself am not sure whether impls in macros should be indented. Just a bit of an eyesore.

Reply:
I think I stole this from the `Arbitrary` implementation. I'm somewhat ambivalent, but too lazy to change it now.